README.md
GNU Queue load-balancing/batch processing system

Copyright 1998-2000 W. G. Krebs (see file 'COPYING' for terms of GPL license)
<wkrebs@gnu.org> 

GNU Queue is distributed in the hope that it is useful but comes 
with absolutely no warranty; see file `COPYING' for details.

INTRODUCTION


This load-balancing system features an innovative 'stub daemon'
mechanism which allows users to control their remote jobs in a nearly
seamless and transparent fashion within a cluster of homogeneous
UNIX workstations. When an interactive remote job is
launched, such as say Matlab, or EMACS interfacing Allegro Lisp, a
stub daemon runs on the remote end. By sending signals to the remote
stub - including hitting the suspend key - the process on the remote
end may be controlled. Resuming the stub resumes the remote job. The
user's environment is almost completely replicated, including not only
environmental variables, but nice values, rlimits, terminal settings
are all replicated on the remote end. Together with MIT_MAGIC_COOKIE_1
(or xhost +; yuck) the system is X-windows transparent as well,
provided the users local DISPLAY variable is set to the fully
qualified pathname of the local machine. 

One of the most appealing features of the stub system even with
experienced users is that asynchronous job control of remote jobs by
the shell is possible and intuitive. One simply runs the stub in the
background under the local shell; the shell notifies the user when the
remote job has a change in status by monitoring the stub daemon. 

When the remote process has terminated, the stub returns the exit
value to the shell; otherwise, the stub simulates a death by the same
signal as that which terminated or suspended the remote job. In this
way, control of the remote process is intuitive even to novice users,
as it is just like controlling a local job from the shell. Many of my
original users had to be reminded that their jobs were, in fact,
running remotely. 

In addition, queue also features a more traditional distributed batch
processing environment, with results returned to the user via
email. In addition, traditional batch processing limitations may be
placed on jobs running in either environment (stub or with the email
mechanism) such as suspension of jobs if the system exceeds a certain
load average, limits on CPU time, disk free requirements, limits on
the times in which jobs may run, etc. (These are documented in the
sample 'profile' file included and in the manuals in `doc/' as well.) 

This version of GNU Queue features (we hope) a nearly
blinking-lights automatic installation process described in
the `INSTALL' file.

The major decision to make is whether or not you wish to install
GNU Queue cluster-wide for all users (requires root privileges
and an NFS shared directory that is root-writable throughout the cluster)
or simply for a single user (which requires no special privileges but does
requires an NFS shared directory writable by that user, something
typical in a homogenuous cluster.)

This decision affects whether or not you supply `./configure' the
--enable-root option. The rest of the installation decisions
are normally taken care of by `./confiugre', see `INSTALL' for further
details.

For additional information please refer to the manuals in `doc/'.

PLEASE SEND US FEEDBACK ON QUEUE!

Whether you have a queue-tip, are queue-less about how to solve a
problem, or simply have another bad queue joke, que[ue] us in at
bug-queue@gnu.org and we'll take our que[ue] from you on how best
to improve the software and documentation.

We have set up a developing mailing list, 'queue-developers.'
Currently, the list is working on several fun projects, including MPI
& PVM support, secure socket connections, AFS & Kerberos support. We're also
porting and improving a nifty utility that lets you monitor and control 
the execution of Queue jobs throughout your cluster. The list is a great way 
to tap into the group's expertise and keep up with the latest developments.

So, come join the fun and keep up with the latest developments by visiting
http://lists.sourceforge.net/mailman/listinfo/queue-developers .

Be sure to visit the application's homepage, 
http://www.gnuqueue.org .





NEWS
*Feb 5, 1999 	W. G. Krebs <wkrebs@gnu.org>

	Version 1.12

	Added check to ReadHosts to make sure that HostFile define was properly
	configured.

	Added MAXHOSTNAMELEN to queue.h on platforms where it is not defined.

	Fixed problem with h_aliases/h_name in canonicalname that was pointed
	out by and patched by a large number of people. The new code
	now counts the number of dots in hostname, h_name, and h_aliases
	and returns the first in that order with the most dots in it.

	Changed address lookup code so that it always vets h_name through
	canonicalhost.

	Eliminated a number of compiler warnings in queued.c by changing
	functions to static. Also, if name lookup fails on incoming socket, 
	string is set to "unresolved" rather than bogus IP addr pointer,
	which eliminated another warning. Compiler warnings proved themselves
	harmless in past versions, but distracted a few users trying to
	track down real bugs on their platform.

*January 28, 1999 	W. G. Krebs <wkrebs@gnu.org>

	Version 1.11 patchlevel 6 Jan 1999

	Made change to queued
	suggested by Arthur Castonguay <arthurc@cadabratech.com> to prevent
	segv on some platforms when looking at null onlyhost string.

*January 27, 1999 	W. G. Krebs <wkrebs@gnu.org>

	Version 1.11 patchlevel 5 Jan 1999

	Fixed the implementation of the error mail shutdown so that
	it actually shuts down the error mail. :)

*January 27, 1999 	W. G. Krebs <wkrebs@gnu.org>

	Version 1.11 patchlevel 4 1999

	Added autoconf code to autodetect solaris

	Added counter to error mail; shuts down system error email after 
	more than 10 error messages go out. Someone might want to modify
	this so that the counter is reset after 24 hours have elapsed.

*January 26, 1999 	W. G. Krebs <wkrebs@gnu.org>

	Version 1.11 patchlevel 3 1999

	Added sigblock.c and made sys_siglist always static when undeclared
	in the never ending effort to support all versions of Solaris.

	Got rid of varargs.h as it hasn't been used in some time.

*January 25, 1999 	W. G. Krebs <wkrebs@gnu.org>

        Version 1.11 patchlevel 2 1999

	Added FreeBSD support by placing sys/time.h at top of define.h and
	defining XCASE and IUCLC as zero on platforms where the includes
	don't define them.

	Added check for missing command line parameters in Queue.

	Changed error to warning if hostname in qhostsfile fails to
	resolve in qlib.c
	
	Fixed bug in sigsetmask in queue.h



queue.h
/*WGK 1998/08/05 */
#include "config.h"
#include "define.h"
#include "queue_endian.h"
#ifndef _QUEUE

#define _QUEUE

/*-> ONLY VERY RARELY WILL YOU NEED TO CHANGE THINGS BELOW THIS POINT <-*/

/*You shouldn't need to mess with these either; they are explained in
queued.c*/

#define QUERY_PORT 1423
#define WAKEUP_PORT 1421
#define SERV_PORT 4700

/*->NO LOCAL CONFIGURATION BELOW THIS POINT<-*/

/*Should be set to no less than what the OS supports. 1024 for
  GNU/Linux???*/

#define MAX_PATH_SIZE 10000

#ifndef _MAXNAMLEN
#define _MAXNAMLEN 256
/*This is the maximum length of a filename under a modern Unix
  filesystem, currently around 255.*/
#endif

#ifndef MAXHOSTNAMELEN
#define MAXHOSTNAMELEN 64
#endif

struct wkstruct{
  char *host;
  float load;
};


#ifdef sethosts
struct wkstruct *Hosts = NULL;
int NHosts;
#else
extern struct wkstruct *Hosts;
extern int NHosts;
#endif

extern char *canonicalhost(char *host);
extern char *localhost(char *host);

extern char *mymalloc(int n);



typedef double DOUBLE;

#undef DIRSIZ_Q
#define	DIRSIZ_Q	32	/* Used to be 14; could be MAXNAMLEN... -IAN! */

typedef int fdev; /*should normally be OK.*/

/*typedef dev_t fdev;*/


#ifndef MAXPATHLEN
#define MAXPATHLEN 1024
#endif

#ifndef MAXBSIZE
#define MAXBSIZE	4096	/* correct for IRIS */
#endif

#ifndef RLIM_NLIMITS
#define RLIM_NLIMITS	6	/* Number of resources, eg RLIMIT_CPU */
#endif

#ifndef R_OK
#define	R_OK	04
#endif

#ifndef SYS_SIGLIST_DECLARED
static char	*sys_siglist[NSIG] = {
  "Signal 0",
  "Hangup",			/* SIGHUP */
  "Interrupt",			/* SIGINT */
  "Quit",			/* SIGQUIT */
  "Illegal instruction",	/* SIGILL */
  "Trace/BPT trap",		/* SIGTRAP */
  "IOT trap",			/* SIGIOT */
  "EMT trap",			/* SIGEMT */
  "Floating point exception",	/* SIGjpE */
  "Killed",			/* SIGKILL */
  "Bus error",			/* SIGBUS */
  "Segmentation fault",		/* SIGSEGV */
  "Bad system call",		/* SIGSYS */
  "Broken pipe",		/* SIGPIPE */
  "Alarm clock",		/* SIGALRM */
  "Terminated",			/* SIGTERM */
  "User-defined signal 1",	/* SIGUSR1 */
  "User-defined signal 2",	/* SIGUSR2 */
  "Child exited",		/* SIGCLD */
  "Power fail",			/* SIGPWR */
  "Signal 20",
  "Signal 21",
  "Signal 22",
  "Signal 23",
  "Signal 24",
  "Signal 25",
  "Signal 26",
  "Signal 27",
  "Signal 28",
  "Signal 29",
  "Signal 30",
#ifndef __hp9000s800
  "Signal 31"
#endif
};
#endif /*SYS_SIGLIST_DECLARED*/


#define	Q_QUEUESTAT		"queuestat"	/* queue status file */
#define	Q_PROFILE		"profile"	/* name of prifile file */
#define	Q_CFDIR			"CFDIR"		/* directory for cf* files */

#define DEV_NULL		"/dev/null"	/* null q_supervisor */

#define Q_STARTNEW	0	/* "exec on" in profile; start new jobs */
#define Q_DRAINING	1	/* "exec off" or "exec drain" in profile */
#define Q_MIGRATING     1       /* "migrate jobs out.*/

#define Q_NOCONTACT 0 /*Don't actually wake up any hosts; just query them.*/
#define Q_CONTACT 1  /*Contact the best host.*/

#define USER_CHECKPOINT 2
#define KERNEL_CHECKPOINT 1
#define NO_CHECKPOINT 0

#define RESTART 1
#define NO_RESTART 0


/* the length of the status prefix */
#define BAT_PREFIX_LEN          2

/* the maximum length of a user-id as defined by utmp */
#define BAT_MAXUID sizeof(((struct utmp *)NULL)->ut_name)

/* startup configuration for queues */
#define BAT_INICE	0	/* initial nice value */
#define BAT_IMAXEXEC	1	/* initial number of running jobs */
#define BAT_ILOADSCHED	25	/* initial load average to schedule jobs */
#define BAT_ILOADSTOP	50	/* initial load average to stop running jobs */
#define BAT_IMINFREE	0	/* initial min. number of free devices */
#define BAT_IRESTARTMODE NO_RESTART
#define BAT_ICHECKPOINTMODE NO_CHECKPOINT
#define BAT_IRESTARTFLAG ""
#define BAT_ICHECKPOINT 26    /*Default load to start checkpoint migrating.*/

#define STRSIZ	256		/* sprintf string buffers */

typedef char bat_Mail;
#define MAIL_START	(1<<0)
#define MAIL_END	(1<<1)
#define MAIL_CRASH	(1<<2)

#define CONTROL_STR     "# BATCH: Start of User Input\n"
#define RETURN_LINES	5

/*
 * Pathnames
 */
/* The next two are defined in Makefile */

#define DEBUGFILE	"/tmp/queued.debug"	/* debug log file */
#define SLEEPTIME	120			/* idle timer between jobs */

/* Old stuff when handle() ran as a QUEUED executable.
#define QUEUED "/home/wkrebs/queue/queue-test/queued"*/



#ifdef SIGSTOP
#define JOBCTL 1		/* Berkeley-style job control available */
#endif /*SIGSTOP*/

#ifdef __hpux
#define POSIX
#endif


#ifdef HAVE_DIRENT_H
#define struct_dir struct dirent
#else
#define struct_dir struct direct
#endif

#define error1(fmt,a1)		{ sprintf(errstr,fmt,a1); error(errstr); }
#define error2(fmt,a1,a2)	{ sprintf(errstr,fmt,a1,a2); error(errstr); }

#define syserror1(fmt,a1)	{ sprintf(errstr,fmt,a1); syserror(errstr); }

/*WGK 1999/01/19  The following defines are needed for SUN-OS 5.x*/

#ifndef sigmask
#define sigmask(sig) \
  (((unsigned long) 1) << (((sig) - 1) % (8 * sizeof (unsigned long int))))
#endif /*sigmask*/

#ifndef HAVE_HERROR
/*WGK Not sure if this is correct, but it will cause it compile, and
  since it is only used once as an error diagnostic, it is not
  essential that this work.*/
#define herror perror
#endif /*HAVE_HERROR*/
/*WGK End SUN-OS stuff.*/

/*WGK 1999/01/24 FreeBSD support stuff*/
/*These constants are Or'd in, so if the platform
  doesn't support them, set them to zero.*/

#ifndef XCASE
#define XCASE   0
#endif /* XCASE*/

#ifndef IUCLC
#define IUCLC 0
#endif /*IUCLC*/

/*WGK End FreeBSD stuff*/

unsigned char *md5(unsigned charstrin, int len);

#endif /*_QUEUE*/
